packages/template-retail-react-app/app/components/bonus-product-view-modal/index.jsx

```diff
@@ -0,0 +1,391 @@
+import {
+    findAvailableBonusDiscountLineItemIds,
+    getRemainingAvailableBonusProductsForProduct,
+    getBonusProductCountsForPromotion
+} from '@salesforce/retail-react-app/app/utils/bonus-product'
+import useNavigation from '@salesforce/retail-react-app/app/hooks/use-navigation'
+import {productViewModalTheme} from '@salesforce/retail-react-app/app/theme/components/project/product-view-modal'
+import {bonusProductViewModalTheme} from '@salesforce/retail-react-app/app/theme/components/project/bonus-product-view-modal'
+import {useToast} from '@salesforce/retail-react-app/app/hooks/use-toast'
+
+/**
+ * A Modal that contains Bonus Product View
+ */
+const BonusProductViewModal = ({
+    product,
+    isOpen,
+    onClose,
+    bonusDiscountLineItemId,
+    promotionId,
+    onReturnToSelection,
+    ...props
+}) => {
+    // Ensure a safe product shape for the modal hook
+    const safeProduct = useMemo(() => {
+        if (!product) return {productId: undefined, variants: [], variationAttributes: []}
+        const id = product.productId || product.id
+        return {
+            productId: id,
+            id,
+            variants: product.variants || [],
+            variationAttributes: product.variationAttributes || [],
+            imageGroups: product.imageGroups || [],
+            type: product.type || {set: false, bundle: false},
+            price: product.price,
+            name: product.name || product.productName
+        }
+    }, [product])
+
+    const productViewModalData = useProductViewModal(safeProduct)
+    const {addItemToNewOrExistingBasket} = useShopperBasketsMutationHelper()
+    const {data: basket} = useCurrentBasket()
+    const navigate = useNavigation()
+
+    const intl = useIntl()
+    const {formatMessage} = intl
+    const showToast = useToast()
+
+    // Calculate bonus counts using promotionId and utility method
+    const {selectedBonusItems: finalSelectedBonusItems, maxBonusItems: finalMaxBonusItems} =
+        useMemo(() => {
+            return getBonusProductCountsForPromotion(basket, promotionId)
+        }, [basket, promotionId])
+
+    const messages = useMemo(
+        () => ({
+            modalLabel: formatMessage(
+                {
+                    id: 'bonus_product_view_modal.modal_label',
+                    defaultMessage: 'Bonus product selection modal for {productName}'
+                },
+                {productName: productViewModalData?.product?.name}
+            ),
+            viewCart: formatMessage({
+                id: 'bonus_product_view_modal.button.view_cart',
+                defaultMessage: 'View Cart'
+            }),
+            backToSelection: formatMessage({
+                id: 'bonus_product_view_modal.button.back_to_selection',
+                defaultMessage: '← Back to Selection'
+            })
+        }),
+        [intl]
+    )
+
+    // Helper function to calculate remaining bonus quantity
```

Comment:
Consider moving all the helper functions to a custom hook or a utils.js file inside the component folder to keep the actual component footprint smaller and more maintainable.